### PR TITLE
Adds binding interface functionality to Budibase

### PR DESCRIPTION
**Pull request by kevmodrome**
## Description
This PR adds binding functionality to props that have the input control.

## Screenshots
https://cln.sh/WozNCS

Not 100% sure of the design as of now. Any suggestions appreciated and welcome.

### CODE REVIEW

File: packages/builder/src/builderStore/fetchBindableProperties.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
+import { cloneDeep, difference } from "lodash/fp"
+
+/**
+ * parameter for fetchBindableProperties function
+ * @typedef {Object} fetchBindablePropertiesParameter
+ * @property {string}  componentInstanceId - an _id of a component that has been added to a screen, whihc you want to fetch bindable props for
+ * @propperty {Object} screen - current screen - where componentInstanceId lives
```
Comment (shogunpurple) [inline packages/builder/src/builderStore/fetchBindableProperties.js:7]: NAB: few typos here

File: packages/builder/src/builderStore/fetchBindableProperties.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    type: "instance",
+    instance: i.instance,
+    // how the binding expression persists, and is used in the app at runtime
+    runtimeBinding: `${contextParentPath}${i.instance._id}.${i.prop}`,
+    // how the binding exressions looks to the user of the builder
+    readableBinding: `${i.instance._instanceName}`,
+  }
+}
+
+const contextToBindables = walkResult => c => {
```
Comment (shogunpurple) [inline packages/builder/src/builderStore/fetchBindableProperties.js:72]: NAB: please try and avoid using single letter variable names, where possible. It harms readability a bit

File: packages/builder/src/builderStore/fetchBindableProperties.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  if (instance._id === targetId) {
+    // found it
+    result.target = instance
+  } else {
+    if (component && component.bindable) {
+      // pushing all components in here initially
+      // but this will not be correct, as some of
+      // these components will be in another context
+      // but we dont know this until the end of the walk
+      // so we will filter in another metod
```
Comment (shogunpurple) [inline packages/builder/src/builderStore/fetchBindableProperties.js:123]: NAB: typo

File: packages/builder/src/components/userInterface/PropertyControl.svelte [hunk 4]
```diff
@@ -8,6 +14,45 @@
...
+  }
+
+  async function replaceBindings(val) {
+    getBindableProperties()
+    // Find all instances of mustasche
+    const boundValues = val.match(/{{([^}]+)}}/g)
+
+    // Replace with names:
+    boundValues &&
+      boundValues.forEach(v => {
```
Comment (shogunpurple) [inline packages/builder/src/components/userInterface/PropertyControl.svelte:45]: NAB: Same thing with single letter variable names, and generic identifiers like `val` - it's good to be more explicit

File: packages/builder/src/components/userInterface/PropertyControl.svelte [hunk 5]
```diff
@@ -17,21 +62,35 @@
...
       }
     }
-    onChange(key, innerVal)
+    replaceBindings(innerVal)
   }
 
   const safeValue = () => {
+    getBindableProperties()
+    let temp = value
+    const boundValues = (value && value.match(/{{([^}]+)}}/g)) || []
```
Comment (shogunpurple) [inline packages/builder/src/components/userInterface/PropertyControl.svelte:71]: NAB: possibly worth moving these regex definitions out into a constant so it's easy to see what it's matching for at a glance

```
const CAPTURE_VAR_INSIDE_MUSTACHE = /{{([^}]+)}}/g
```

File: packages/client/src/state/store.js [hunk 6]
```diff
@@ -1,9 +1,104 @@
...
+}
+
+const subscribe = (subscription, storeKey) => {
+  if (!storeKey) {
+    return rootStore.subscribe(subscription)
+  }
+  const contextStore = contextStores[storeKey]
+
+  // we are subscribing to multiple stores,
+  // we dont want to each subscription the first time
```
Comment (shogunpurple) [inline packages/client/src/state/store.js:43]: NAB: this sentence doesn't make any sense

**Review by shogunpurple — APPROVED** [review]
Logic looks good - Just some style related stuff for readability.

**Comment by kevmodrome** [comment]
This should now be ready to be merged 😲

**Comment by kevmodrome** [comment]
I'm going to go ahead and merge this now, I couldn't figure out the cypress issue and will raise an issue about it

**Pull request merged**